feat!: cleanup contract.at and registerContract

Thunkar opened the pull request:
Closes: https://linear.app/aztec-labs/issue/F-73/refactor-registercontract, https://linear.app/aztec-labs/issue/F-134/cannot-registercontract-and-simulatesend-on-the-same-walletbatch-using

This PR greatly simplifies `registerContract` on the wallet interface, while at the same time makes the registering flow more explicit. In the past we were just spamming registrations, which is a no-go for real wallets due to the size of artifacts. 

Furthermore, it simplifies our contract interfaces by removing the instance from them. They were previously always retrieved from the wallet (again, very costly) while instances are only required:

* When an app first registers a contract
* When an app tries to update a contract

If an app needs a contract instance, it can use any of these methods (which it was inadvertedly doing before!)
* Generate it via `getContractInstanceFromInstantiationParams`
* Get it from the node provided the instance is published
* Get it from the result of `Contract.deploy().send().wait()` (if you deploy you know the instance already!)

mverzilli commented on yarn-project/aztec.js/src/contract/contract.ts:
```diff
@@ -45,7 +46,7 @@ export class Contract extends ContractBase {
    * @param artifact - Build artifact of the contract.
    * @param args - Arguments for the constructor.
-   * @param constructorName - The name of the constructor function to call.
+   * @param constructorName - The name of thse constructor function to call.
```
typo

```suggestion
   * @param constructorName - The name of the constructor function to call.
```

mverzilli reviewed (approved):
Nice!

Thunkar committed:
feat!: cleanup contract.at and registerContract

Closes: https://linear.app/aztec-labs/issue/F-73/refactor-registercontract, https://linear.app/aztec-labs/issue/F-134/cannot-registercontract-and-simulatesend-on-the-same-walletbatch-using

This PR greatly simplifies `registerContract` on the wallet interface, while at the same time makes the registering flow more explicit. In the past we were just spamming registrations, which is a no-go for real wallets due to the size of artifacts.

Furthermore, it simplifies our contract interfaces by removing the instance from them. They were previously always retrieved from the wallet (again, very costly) while instances are only required:

* When an app first registers a contract
* When an app tries to update a contract

If an app needs a contract instance, it can use any of these methods (which it was inadvertedly doing before!)
* Generate it via `getContractInstanceFromInstantiationParams`
* Get it from the node provided the instance is published
* Get it from the result of `Contract.deploy().send().wait()` (if you deploy you know the instance already!)

Co-authored-by: Gregorio Juliana <gregojquiros@gmail.com>
Co-authored-by: thunkar <gregojquiros@gmail.com>
